File: partiql-plan/src/main/java/org/partiql/plan/rex/RexArray.java
package org.partiql.plan.rex;

import org.jetbrains.annotations.NotNull;
import org.partiql.plan.Operand;
import org.partiql.plan.Operator;
import org.partiql.plan.OperatorVisitor;
import org.partiql.types.PType;

import java.util.List;

/**
 * Logical array expression abstract base class.
 */
public abstract class RexArray extends RexBase {

    /**
     * @return new RexArray instance
     */
    @NotNull
    public static RexArray create(@NotNull List<Rex> values) {
        return new Impl(values);
    }

    /**
     * @return the values of the array, also the operands.
     */
    @NotNull
    public abstract List<Rex> getValues();

    @NotNull
    @Override
    protected final RexType type() {
        return RexType.of(PType.array());
    }

    @NotNull
    @Override
    protected final List<Operand> operands() {
        Operand c0 = Operand.vararg(getValues());
        return List.of(c0);
    }

    @Override
    public <R, C> R accept(@NotNull OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitArray(this, ctx);
    }

    private static class Impl extends RexArray {

        private final List<Rex> values;

        private Impl(List<Rex> values) {
            this.values = values;
        }

        @NotNull
        @Override
        public List<Rex> getValues() {
            return values;
        }
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/rex/RexArray.kt
package org.partiql.plan.rex

/**
 * TODO DOCUMENTATION
 */
public interface RexArray : Rex {

    public fun getValues(): Collection<Rex>

    override fun getChildren(): Collection<Rex> = getValues().toList()

    override fun <R, C> accept(visitor: RexVisitor<R, C>, ctx: C): R = visitor.visitArray(this, ctx)
}

/**
 * Default [RexArray] operator for extension.
 */
internal class RexArrayImpl(values: Collection<Rex>, type: RexType) : RexArray {

    // DO NOT USE FINAL
    private var _values = values
    private var _type = type

    override fun getValues(): Collection<Rex> = _values

    override fun getChildren(): Collection<Rex> = _values.toList()

    override fun getType(): RexType = _type

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RexArray) return false
        if (_values != other.getValues()) return false
        return true
    }

    override fun hashCode(): Int = _values.hashCode()
}
